packages/react-router/lib/router/utils.ts

```diff
@@ -611,28 +611,53 @@ type RegexMatchPlus<
     : never
   : never;
 
-// Recursive helper for finding path parameters in the absence of wildcards
-type _PathParam<Path extends string> =
+// Union of all path parameters - optional and required
+export type PathParam<Path extends string> =
+  | RequiredPathParam<Path>
+  | OptionalPathParam<Path>;
+
+// Recursive helper for finding required path parameters in the absence of wildcards
+type _RequiredPathParam<Path extends string> =
+  // split path into individual path segments
+  Path extends `${infer L}/${infer R}`
+    ? _RequiredPathParam<L> | _RequiredPathParam<R>
+    : // Ignore optional params
+      Path extends `:${string}?${string}`
+      ? never
+      : // Match required params
+        Path extends `:${infer Param}`
+        ? RegexMatchPlus<ParamChar, Param>
+        : never;
+
+type RequiredPathParam<Path extends string> =
+  // check if path is just a wildcard
+  Path extends "*" | "/*"
+    ? "*"
+    : // look for wildcard at the end of the path
+      Path extends `${infer Rest}/*`
+      ? "*" | _RequiredPathParam<Rest>
+      : // look for params in the absence of wildcards
+        _RequiredPathParam<Path>;
```

Comment:
This is identical to `OptionalPathParam` below except in the recursive helper it calls.  I tried to share them and take the helper as a generic but couldn't figure out how - if that's possible we could DRY it up a bit:

```ts
export type PathParam<Path extends string> =
  | _PathParam<Path, _RequiredPathParam>
  | _PathParam<Path, _OptionalPathParam>
```